api/src/pcapi/core/educational/models.py

```diff
@@ -183,6 +404,268 @@ def mark_as_refused(self) -> None:
+        AND (
+            (SELECT "quantity" FROM stock WHERE id=NEW."stockId")
+            <
+            (SELECT SUM(quantity) FROM booking WHERE "stockId"=NEW."stockId" AND status != '{BookingStatus.CANCELLED.value}')
+            )
+        THEN RAISE EXCEPTION 'tooManyBookings'
+                    USING HINT = 'Number of bookings cannot exceed "stock.quantity"';
+    END IF;
+
+    IF (
+        (NEW."educationalBookingId" IS NULL AND OLD."educationalBookingId" IS NULL)
+        AND (NEW."individualBookingId" IS NOT NULL OR OLD."individualBookingId" IS NOT NULL)
+        AND (
+        -- If this is a new booking, we probably want to check the wallet.
+        OLD IS NULL
+        -- If we're updating an existing booking...
+        OR (
+            -- Check the wallet if we are changing the quantity or the amount
+            -- The backend should never do that, but let's be defensive.
+            (NEW."quantity" != OLD."quantity" OR NEW."amount" != OLD."amount")
+            -- If amount and quantity are unchanged, we want to check the wallet
+            -- only if we are UNcancelling a booking. (Users with no credits left
+            -- should be able to cancel their booking. Also, their booking can
+            -- be marked as used or not used.)
+            OR (NEW.status != OLD.status AND OLD.status = '{BookingStatus.CANCELLED.value}' AND NEW.status != '{BookingStatus.CANCELLED.value}')
+        )
+        )
+        AND (
+            -- Allow to book free offers even with no credit left (or expired deposits)
+            (deposit_id IS NULL AND NEW."amount" != 0)
+            OR (deposit_id IS NOT NULL AND get_deposit_balance(deposit_id, false) < 0)
+        )
+    )
+    THEN RAISE EXCEPTION 'insufficientFunds'
+                USING HINT = 'The user does not have enough credit to book';
+    END IF;
+
+    RETURN NEW;
+    END;
+    $$ LANGUAGE plpgsql;
+
+    DROP TRIGGER IF EXISTS booking_update ON booking;
+    CREATE CONSTRAINT TRIGGER booking_update
+    AFTER INSERT
+    OR UPDATE OF quantity, amount, status, "userId"
+    ON booking
+    FOR EACH ROW EXECUTE PROCEDURE check_booking()
+    """
+event.listen(CollectiveBooking.__table__, "after_create", DDL(CollectiveBooking.trig_ddl))
+
+CollectiveBooking.trig_update_cancellationDate_on_isCancelled_ddl = f"""
+    CREATE OR REPLACE FUNCTION save_cancellation_date()
+    RETURNS TRIGGER AS $$
+    BEGIN
+        IF NEW.status = '{BookingStatus.CANCELLED.value}' AND OLD."cancellationDate" IS NULL AND NEW."cancellationDate" THEN
+            NEW."cancellationDate" = NOW();
+        ELSIF NEW.status != '{BookingStatus.CANCELLED.value}' THEN
+            NEW."cancellationDate" = NULL;
+        END IF;
+        RETURN NEW;
+    END;
+    $$ LANGUAGE plpgsql;
+
+    DROP TRIGGER IF EXISTS stock_update_cancellation_date ON booking;
+
+    CREATE TRIGGER stock_update_cancellation_date
+    BEFORE INSERT OR UPDATE OF status ON booking
```

Comment:
copier-coller trop rapide ! :)

```suggestion
    BEFORE INSERT OR UPDATE OF status ON collective_booking
```

Reply:
À reprendre en effet, pour l'instant pas généré par migration